File: 351002/Khmel/Task320/Services/StoryService.cs
public class StoryService : IStoryService
{
    private readonly IRepository<Story> _repository;
    private readonly IRepository<Writer> _writerRepository;

    public StoryService(IRepository<Story> repository, IRepository<Writer> writerRepository)
    {
        _repository = repository;
        _writerRepository = writerRepository;
    }

    public StoryResponseTo GetById(long id)
    {
        var story = _repository.GetById(id);
        if (story == null)
            throw new KeyNotFoundException($"Story с id={id} не найдена");
        return ToResponse(story);
    }

    public IEnumerable<StoryResponseTo> GetAll() =>
        _repository.GetAll().Select(ToResponse);

    public StoryResponseTo Create(StoryRequestTo request)
    {
        var writer = _writerRepository.GetById(request.WriterId);
        if (writer == null)
            throw new KeyNotFoundException($"Writer с id={request.WriterId} не найден");

        var duplicate = _repository.GetAll()
            .FirstOrDefault(s => s.Title == request.Title);
        if (duplicate != null)
            throw new ForbiddenException($"Story с title={request.Title} уже существует");

        var story = ToModel(request);
        return ToResponse(_repository.Create(story));
    }

    public StoryResponseTo Update(StoryRequestTo request)
    {
        var story = ToModel(request);
        return ToResponse(_repository.Update(story));
    }

    public void Delete(long id)
    {
        if (!_repository.Delete(id))
            throw new KeyNotFoundException($"Story с id={id} не найдена");
    }

    private Story ToModel(StoryRequestTo dto) => new Story
    {
        Id = dto.Id,
        WriterId = dto.WriterId,
        Title = dto.Title,
        Content = dto.Content
    };

    private StoryResponseTo ToResponse(Story model) => new StoryResponseTo
    {
        Id = model.Id,
        WriterId = model.WriterId,
        Title = model.Title,
        Content = model.Content,
        Created = model.Created,
        Modified = model.Modified
    };
}

File: 351004/Bunevich/labs/task330/src/main/java/com/example/demo/servises/StoryService.java
package com.example.demo.servises;

import com.example.demo.dto.request.StoryRequestTo;
import com.example.demo.dto.response.StoryResponseTo;
import com.example.demo.exeptionHandler.ConflictException;
import com.example.demo.exeptionHandler.ResourceNotFoundException;
import com.example.demo.mapper.StoryMapper;
import com.example.demo.models.Story;
import com.example.demo.models.Tag;
import com.example.demo.models.Writer;
import com.example.demo.repository.StoryRepository;
import com.example.demo.repository.TagRepository;
import com.example.demo.repository.WriterRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Service
public class StoryService {
    private final StoryMapper storyMapper;
    private final StoryRepository storyRepository;
    private final WriterRepository writerRepository;
    private final TagRepository tagRepository;

    public StoryService(StoryMapper storyMapper,
                        StoryRepository storyRepository,
                        WriterRepository writerRepository,
                        TagRepository tagRepository) {
        this.storyMapper = storyMapper;
        this.storyRepository = storyRepository;
        this.writerRepository = writerRepository;
        this.tagRepository = tagRepository;
    }

    public List<StoryResponseTo> findAll(int page, int size, String sortBy, String sortDir, String title){
        Sort sort = "desc".equalsIgnoreCase(sortDir) ? Sort.by(sortBy).descending() : Sort.by(sortBy).ascending();
        Specification<Story> spec = (root, query, cb) ->
                title == null || title.isBlank() ? cb.conjunction() : cb.like(cb.lower(root.get("title")), "%" + title.toLowerCase() + "%");
         return storyMapper.storyListToResponseList(storyRepository.findAll(spec, PageRequest.of(page, size, sort)).getContent());
    }

    public StoryResponseTo create(StoryRequestTo storyRequestTo){
        validateStoryContent(storyRequestTo.content);
        Writer writer = writerRepository.findById(storyRequestTo.writerId)
                .orElseThrow(() -> new ResourceNotFoundException("Writer", storyRequestTo.writerId));
        Story story = storyMapper.toEntity(storyRequestTo);
        story.setWriter(writer);
        story.setCreated(LocalDateTime.now());
        story.setModified(LocalDateTime.now());
        if (storyRequestTo.getTagIds() != null && !storyRequestTo.getTagIds().isEmpty()) {
            Set<Tag> tags = new HashSet<>(tagRepository.findAllById(storyRequestTo.getTagIds()));
            story.setTags(tags);
        }
        Story savedStory = storyRepository.save(story);
        return storyMapper.toResponse(savedStory);
    }

    public void delete(Long id){
        if(!storyRepository.existsById(id)){
            throw new ResourceNotFoundException("Story", id);
        }
        storyRepository.deleteById(id);
    }

    public StoryResponseTo findById(Long id){
        Story story = storyRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Story", id));
        return storyMapper.toResponse(story);
    }

    public StoryResponseTo update(Long id, StoryRequestTo storyRequestTo){
        validateStoryContent(storyRequestTo.content);
        Story existing = storyRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Story", id));
        Writer writer = writerRepository.findById(storyRequestTo.writerId)
                .orElseThrow(() -> new ResourceNotFoundException("Writer", storyRequestTo.writerId));
        Story story = storyMapper.toEntity(storyRequestTo);
        story.setId(existing.getId());
        story.setWriter(writer);
        story.setCreated(existing.getCreated());
        story.setModified(LocalDateTime.now());
        if (storyRequestTo.getTagIds() != null) {
            Set<Tag> tags = new HashSet<>(tagRepository.findAllById(storyRequestTo.getTagIds()));
            story.setTags(tags);
        } else {
            story.setTags(existing.getTags());
        }
        return storyMapper.toResponse(storyRepository.save(story));
    }

    private void validateStoryContent(String content) {
        if (content != null && content.startsWith("other-content")) {
            throw new ConflictException("Story content is forbidden");
        }
    }

}
